File: python_programs_and_containers/basic_container_lite/libraries/file_loader/test_files/kb_go/postgres/kb_construct/kb_construct_module/ConstructRPCClientTable.go
package kb_construct_module

import (
	"database/sql"
	"encoding/json"
	"fmt"

	"github.com/google/uuid"
	_ "github.com/lib/pq"
)

// ConstructRPCClientTable manages RPC client table operations
type ConstructRPCClientTable struct {
	conn        *sql.DB
	constructKB *ConstructKB
	database    string
	tableName   string
}

// RPCClientFieldResult represents the result of adding an RPC client field
type RPCClientFieldResult struct {
	RPCClient  string                 `json:"rpc_client"`
	Message    string                 `json:"message"`
	Properties map[string]interface{} `json:"properties"`
	Data       string                 `json:"data"`
}

// QueueAdjustmentClientResult represents the result of queue adjustment for a client path
type QueueAdjustmentClientResult struct {
	Added   int    `json:"added"`
	Removed int    `json:"removed"`
	Error   string `json:"error,omitempty"`
}

// NewConstructRPCClientTable creates a new instance of ConstructRPCClientTable
func NewConstructRPCClientTable(conn *sql.DB, constructKB *ConstructKB, database string) (*ConstructRPCClientTable, error) {
	crt := &ConstructRPCClientTable{
		conn:        conn,
		constructKB: constructKB,
		database:    database,
		tableName:   database + "_rpc_client",
	}

	if err := crt.setupSchema(); err != nil {
		return nil, fmt.Errorf("error setting up schema: %w", err)
	}

	return crt, nil
}

// setupSchema sets up the database schema
func (crt *ConstructRPCClientTable) setupSchema() error {
	// Create ltree extension
	if _, err := crt.conn.Exec("CREATE EXTENSION IF NOT EXISTS ltree;"); err != nil {
		return fmt.Errorf("error creating ltree extension: %w", err)
	}

	// Drop existing table
	dropQuery := fmt.Sprintf("DROP TABLE IF EXISTS %s CASCADE", crt.tableName)
	if _, err := crt.conn.Exec(dropQuery); err != nil {
		return fmt.Errorf("error dropping table: %w", err)
	}

	// Create the RPC client table
	createTableQuery := fmt.Sprintf(`
		CREATE TABLE %s (
			id SERIAL PRIMARY KEY,
			
			-- Reference to the request
			request_id UUID NOT NULL,
			
			-- Path to identify the RPC client queue for routing responses
			client_path ltree NOT NULL,
			server_path ltree NOT NULL,
			
			-- Response information
			transaction_tag TEXT NOT NULL DEFAULT 'none',
			rpc_action TEXT NOT NULL DEFAULT 'none',

			response_payload JSONB NOT NULL,
			response_timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW(), -- UTC timestamp
			
			-- Boolean to identify new/unprocessed results
			is_new_result BOOLEAN NOT NULL DEFAULT FALSE
		);`, crt.tableName)

	if _, err := crt.conn.Exec(createTableQuery); err != nil {
		return fmt.Errorf("error creating table: %w", err)
	}

	fmt.Println("rpc_client table created.")
	return nil
}

// AddRPCClientField adds a new RPC client field to the knowledge base
func (crt *ConstructRPCClientTable) AddRPCClientField(rpcClientKey string, queueDepth int, description string) (*RPCClientFieldResult, error) {
	properties := map[string]interface{}{
		"queue_depth": queueDepth,
	}

	// Add the node to the knowledge base
	if err := crt.constructKB.AddInfoNode("KB_RPC_CLIENT_FIELD", rpcClientKey, properties, map[string]interface{}{}, description); err != nil {
		return nil, fmt.Errorf("error adding info node: %w", err)
	}

	fmt.Printf("Added rpc_client field '%s' with properties: %v\n", rpcClientKey, properties)

	result := &RPCClientFieldResult{
		RPCClient:  "success",
		Message:    fmt.Sprintf("rpc_client field '%s' added successfully", rpcClientKey),
		Properties: properties,
		Data:       description,
	}

	return result, nil
}

// RemoveUnspecifiedEntries removes entries from rpc_client_table where client_path is not in the specified list
func (crt *ConstructRPCClientTable) RemoveUnspecifiedEntries(specifiedClientPaths []string) (int, error) {
	if len(specifiedClientPaths) == 0 {
		fmt.Println("Warning: No client_paths specified. No entries will be removed.")
		return 0, nil
	}

	// Filter out empty strings
	var validPaths []string
	for _, path := range specifiedClientPaths {
		if path != "" {
			validPaths = append(validPaths, path)
		}
	}

	if len(validPaths) == 0 {
		fmt.Println("Warning: No valid client_paths found after filtering. No entries will be removed.")
		return 0, nil
	}

	fmt.Printf("Processing %d valid client paths\n", len(validPaths))

	// Begin transaction
	tx, err := crt.conn.Begin()
	if err != nil {
		return 0, fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	// Create temporary table
	if _, err := tx.Exec("CREATE TEMP TABLE IF NOT EXISTS valid_client_paths (path text)"); err != nil {
		return 0, fmt.Errorf("error creating temp table: %w", err)
	}

	// Clear the temp table
	if _, err := tx.Exec("DELETE FROM valid_client_paths"); err != nil {
		return 0, fmt.Errorf("error clearing temp table: %w", err)
	}

	// Insert paths in batches
	batchSize := 1000
	insertStmt, err := tx.Prepare("INSERT INTO valid_client_paths VALUES ($1)")
	if err != nil {
		return 0, fmt.Errorf("error preparing insert statement: %w", err)
	}
	defer insertStmt.Close()

	for i := 0; i < len(validPaths); i += batchSize {
		end := i + batchSize
		if end > len(validPaths) {
			end = len(validPaths)
		}

		for j := i; j < end; j++ {
			if _, err := insertStmt.Exec(validPaths[j]); err != nil {
				return 0, fmt.Errorf("error inserting path %s: %w", validPaths[j], err)
			}
		}
	}

	// Delete entries not in our temp table
	deleteQuery := fmt.Sprintf(`
		DELETE FROM %s
		WHERE client_path::text NOT IN (
			SELECT path FROM valid_client_paths
		)`, crt.tableName)

	result, err := tx.Exec(deleteQuery)
	if err != nil {
		return 0, fmt.Errorf("error deleting entries: %w", err)
	}

	deletedCount, err := result.RowsAffected()
	if err != nil {
		return 0, fmt.Errorf("error getting rows affected: %w", err)
	}

	// Commit transaction
	if err := tx.Commit(); err != nil {
		return 0, fmt.Errorf("error committing transaction: %w", err)
	}

	// Clean up temp table (best effort)
	crt.conn.Exec("DROP TABLE IF EXISTS valid_client_paths")

	fmt.Printf("Removed %d unspecified entries from %s\n", deletedCount, crt.tableName)
	return int(deletedCount), nil
}

// AdjustQueueLength adjusts the number of records for multiple client paths to match their specified queue lengths
func (crt *ConstructRPCClientTable) AdjustQueueLength(specifiedClientPaths []string, specifiedQueueLengths []int) (map[string]QueueAdjustmentClientResult, error) {
	if len(specifiedClientPaths) != len(specifiedQueueLengths) {
		return nil, fmt.Errorf("the specified_client_paths and specified_queue_lengths lists must be of equal length")
	}

	results := make(map[string]QueueAdjustmentClientResult)

	// Begin transaction
	tx, err := crt.conn.Begin()
	if err != nil {
		return nil, fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	for i, clientPath := range specifiedClientPaths {
		queueLength := specifiedQueueLengths[i]

		if queueLength < 0 {
			results[clientPath] = QueueAdjustmentClientResult{Error: "Invalid queue length (negative)"}
			continue
		}

		// Count current records
		countQuery := fmt.Sprintf(`
			SELECT COUNT(*) 
			FROM %s
			WHERE client_path = $1::ltree`, crt.tableName)

		var currentCount int
		if err := tx.QueryRow(countQuery, clientPath).Scan(&currentCount); err != nil {
			results[clientPath] = QueueAdjustmentClientResult{Error: err.Error()}
			continue
		}

		pathResult := QueueAdjustmentClientResult{Added: 0, Removed: 0}

		// Remove excess records
		if currentCount > queueLength {
			recordsToRemove := currentCount - queueLength
			deleteQuery := fmt.Sprintf(`
				DELETE FROM %s
				WHERE id IN (
					SELECT id
					FROM %s
					WHERE client_path = $1::ltree
					ORDER BY response_timestamp ASC
					LIMIT $2
				)
				RETURNING id`, crt.tableName, crt.tableName)

			rows, err := tx.Query(deleteQuery, clientPath, recordsToRemove)
			if err != nil {
				results[clientPath] = QueueAdjustmentClientResult{Error: err.Error()}
				continue
			}
			defer rows.Close()

			removedCount := 0
			for rows.Next() {
				var id int
				if err := rows.Scan(&id); err == nil {
					removedCount++
				}
			}
			pathResult.Removed = removedCount

		} else if currentCount < queueLength {
			// Add missing records
			recordsToAdd := queueLength - currentCount
			insertQuery := fmt.Sprintf(`
				INSERT INTO %s (
					request_id, client_path, server_path,
					transaction_tag, rpc_action,
					response_payload, response_timestamp, is_new_result
				)
				VALUES ($1, $2::ltree, $3::ltree, $4, $5, $6::jsonb, NOW(), FALSE)`, crt.tableName)

			for j := 0; j < recordsToAdd; j++ {
				requestID := uuid.New().String()
				emptyPayload := "{}"

				if _, err := tx.Exec(insertQuery,
					requestID,
					clientPath,
					clientPath, // default server_path = client_path
					"none",     // default transaction_tag
					"none",     // default rpc_action
					emptyPayload,
				); err != nil {
					results[clientPath] = QueueAdjustmentClientResult{Error: err.Error()}
					break
				}
				pathResult.Added++
			}
		}

		results[clientPath] = pathResult
	}

	// Commit transaction
	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("error committing transaction: %w", err)
	}

	return results, nil
}

// RestoreDefaultValues restores default values for all fields except client_path
func (crt *ConstructRPCClientTable) RestoreDefaultValues() (int, error) {
	updateQuery := fmt.Sprintf(`
		UPDATE %s
		SET 
			request_id = (SELECT gen_random_uuid()),  -- Unique UUID per record
			server_path = client_path,  -- Set server_path to match client_path
			transaction_tag = 'none',
			rpc_action = 'none',
			response_payload = '{}'::jsonb,
			response_timestamp = NOW(),
			is_new_result = FALSE
		RETURNING id`, crt.tableName)

	rows, err := crt.conn.Query(updateQuery)
	if err != nil {
		return 0, fmt.Errorf("error updating records: %w", err)
	}
	defer rows.Close()

	updatedCount := 0
	for rows.Next() {
		var id int
		if err := rows.Scan(&id); err != nil {
			return 0, fmt.Errorf("error scanning id: %w", err)
		}
		updatedCount++
	}

	return updatedCount, nil
}

// CheckInstallation synchronizes the knowledge_base and rpc_client_table
func (crt *ConstructRPCClientTable) CheckInstallation() error {
	// Get specified paths from knowledge_table
	query := fmt.Sprintf(`
		SELECT path, properties FROM %s 
		WHERE label = 'KB_RPC_CLIENT_FIELD';`, crt.database)

	rows, err := crt.conn.Query(query)
	if err != nil {
		return fmt.Errorf("error retrieving knowledge base fields: %w", err)
	}
	defer rows.Close()

	var paths []string
	var lengths []int

	fmt.Println("specified_paths_data:")
	for rows.Next() {
		var path string
		var propertiesJSON []byte

		if err := rows.Scan(&path, &propertiesJSON); err != nil {
			return fmt.Errorf("error scanning row: %w", err)
		}

		var properties map[string]interface{}
		if err := json.Unmarshal(propertiesJSON, &properties); err != nil {
			return fmt.Errorf("error unmarshaling properties: %w", err)
		}

		paths = append(paths, path)

		// Extract queue_depth
		if queueDepth, ok := properties["queue_depth"].(float64); ok {
			lengths = append(lengths, int(queueDepth))
		} else {
			return fmt.Errorf("queue_depth not found or invalid for path %s", path)
		}

		fmt.Printf("  path: %s, properties: %v\n", path, properties)
	}

	// Execute the three operations
	if _, err := crt.RemoveUnspecifiedEntries(paths); err != nil {
		return fmt.Errorf("error removing unspecified entries: %w", err)
	}

	if _, err := crt.AdjustQueueLength(paths, lengths); err != nil {
		return fmt.Errorf("error adjusting queue length: %w", err)
	}

	if _, err := crt.RestoreDefaultValues(); err != nil {
		return fmt.Errorf("error restoring default values: %w", err)
	}

	return nil
}

File: python_programs_and_containers/building_blocks/building_blocks/knowledge_base/construct_kb/construct_rpc_client_table.py
import psycopg2
import json
import uuid
from psycopg2 import sql
from psycopg2.extensions import adapt

class Construct_RPC_Client_Table:
    """
    This class is designed to construct a rpc_client table with header
    and info nodes, using a stack-based approach to manage the path. It also
    manages a connection to a PostgreSQL database and sets up the schema.
    """
    def __init__(self, conn, cursor,construct_kb,database):
        self.conn = conn
        self.cursor = cursor
        self.construct_kb = construct_kb
        self.database = database
        self.table_name = self.database + "_rpc_client"
 # Execute the SQL script to set up the schema
        self._setup_schema()

    def _setup_schema(self):
        """
        Sets up the database schema (tables, functions, etc.).
   
        # Use psycopg2.sql module to construct SQL queries safely. This prevents SQL injection.
        # ltree extension needs to be created.
        """
        create_extensions_script = sql.SQL("""
            CREATE EXTENSION IF NOT EXISTS ltree;
        """)
        query = sql.SQL("DROP TABLE IF EXISTS {table_name} CASCADE").format(
            table_name=sql.Identifier(self.table_name)
        )
        self.cursor.execute(query)
        # Create the knowledge_base table
        create_table_script = sql.SQL("""
             CREATE TABLE   {table_name} (
                id SERIAL PRIMARY KEY,
                
                -- Reference to the request
                request_id UUID NOT NULL,
                
                -- Path to identify the RPC client queue for routing responses
                client_path ltree NOT NULL,
                server_path ltree NOT NULL,
                
                -- Response information
                transaction_tag TEXT NOT NULL DEFAULT 'none',
                rpc_action TEXT NOT NULL DEFAULT 'none',

                response_payload JSONB NOT NULL,
                response_timestamp TIMESTAMPTZ NOT NULL DEFAULT NOW(), -- UTC timestamp
                
                -- Boolean to identify new/unprocessed results
                is_new_result BOOLEAN NOT NULL DEFAULT FALSE
                
              
            );
        """).format(table_name=sql.Identifier(self.table_name))
        self.cursor.execute(create_table_script)
        self.conn.commit()  # Commit the changes
        print("rpc_client table created.")

    def add_rpc_client_field(self, rpc_client_key,queue_depth, description):
        """
        Add a new rpc_client field to the knowledge base
        
        Args:
            rpc_client_key (str): The key/name of the rpc_client field
            description (str): The description of the rpc_client field
            
        Raises:
            TypeError: If rpc_client_key is not a string or initial_properties is not a dictionary
        """
        if not isinstance(rpc_client_key, str):
            raise TypeError("rpc_client_key must be a string")
        if not isinstance(description, str):
            raise TypeError("description must be a string")
        if not isinstance(queue_depth, int):
            raise TypeError("queue_depth must be an integer")
        
        properties = {"queue_depth": queue_depth}
        
        
        # Convert dictionaries to JSON strings
        
        
        # Add the node to the knowledge base
        self.construct_kb.add_info_node("KB_RPC_CLIENT_FIELD", rpc_client_key, properties,{},description)
        
        print(f"Added rpc_client field '{rpc_client_key}' with properties: {properties}")
        
        return {
            "rpc_client": "success",
            "message": f"rpc_client field '{rpc_client_key}' added successfully",
            "properties": properties,
            "data": description
        }
    def remove_unspecified_entries(self, specified_client_paths):
        """
        Remove entries from rpc_client_table where the client_path is not in the specified list,
        handling large lists of paths efficiently.
        
        Args:
            specified_client_paths (list): List of valid client_paths to keep
            
        Returns:
            int: Number of deleted records
        """
        try:
            if not specified_client_paths:
                print("Warning: No client_paths specified. No entries will be removed.")
                return 0
            
            # Filter out None values and convert all paths to strings
            valid_paths = []
            for path in specified_client_paths:
                if path is not None:
                    valid_paths.append(str(path))
            
            if not valid_paths:
                print("Warning: No valid client_paths found after filtering. No entries will be removed.")
                return 0
            
            print(f"Processing {len(valid_paths)} valid client paths")
            
            # Create a temporary table to store valid paths for efficient processing
            # Don't use explicit transaction control to avoid set_session conflicts
            self.cursor.execute("""
                CREATE TEMP TABLE IF NOT EXISTS valid_client_paths (path text)
            """)
            
            # Clear the table in case it already exists
            self.cursor.execute("DELETE FROM valid_client_paths")
            
            # Insert paths in batches to avoid parameter limits
            batch_size = 1000
            for i in range(0, len(valid_paths), batch_size):
                batch = valid_paths[i:i+batch_size]
                args = [(path,) for path in batch]
                try:
                    self.cursor.executemany("""
                        INSERT INTO valid_client_paths VALUES (%s)
                    """, args)
                except Exception as batch_error:
                    print(f"Error inserting batch {i//batch_size + 1}: {batch_error}")
                    print(f"Problematic batch: {batch}")
                    raise
            
            # Delete entries where client_path is not in our temp table
            delete_query = sql.SQL("""
                DELETE FROM {table_name}
                WHERE client_path::text NOT IN (
                    SELECT path FROM valid_client_paths
                )
            """).format(table_name=sql.Identifier(self.table_name))
            
            self.cursor.execute(delete_query)
            deleted_count = self.cursor.rowcount
            
            # Commit the transaction
            self.conn.commit()
            
            # Clean up the temporary table
            try:
                self.cursor.execute("DROP TABLE IF EXISTS valid_client_paths")
            except Exception as cleanup_error:
                print(f"Warning: Could not clean up temporary table: {cleanup_error}")
            
            print(f"Removed {deleted_count} unspecified entries from {self.table_name}")
            return deleted_count
            
        except Exception as e:
            print(f"Error in remove_unspecified_entries: {e}")
            # Roll back in case of error - but don't force autocommit changes
            try:
                if hasattr(self, 'conn') and self.conn:
                    self.conn.rollback()
            except Exception as rollback_error:
                print(f"Warning: Could not rollback transaction: {rollback_error}")
            raise Exception(f"Error in remove_unspecified_entries: {e}")
        
        
    def adjust_queue_length(self, specified_client_paths, specified_queue_lengths):
        """
        Adjust the number of records for multiple client paths to match their specified queue lengths.
        
        Args:
            specified_client_paths (list): List of client paths to adjust
            specified_queue_lengths (list): List of desired queue lengths corresponding to each client path
            
        Returns:
            dict: A dictionary with client paths as keys and operation results as values
        """
        if len(specified_client_paths) != len(specified_queue_lengths):
            raise ValueError("The specified_client_paths and specified_queue_lengths lists must be of equal length")
        
        results = {}
        table_ident = sql.Identifier(*self.table_name.split('.'))  # supports schema-qualified table names
        
        for i, client_path in enumerate(specified_client_paths):
            queue_length = specified_queue_lengths[i]
            
            if queue_length < 0:
                results[client_path] = {"error": "Invalid queue length (negative)"}
                continue
            
            # Count current records
            count_query = sql.SQL("""
                SELECT COUNT(*) 
                FROM {table}
                WHERE client_path = %s::ltree
            """).format(table=table_ident)
            
            self.cursor.execute(count_query, (client_path,))
            current_count = self.cursor.fetchone()[0]
            
            path_result = {'added': 0, 'removed': 0}
            
            # Remove excess records
            if current_count > queue_length:
                records_to_remove = current_count - queue_length
                delete_query = sql.SQL("""
                    DELETE FROM {table}
                    WHERE id IN (
                        SELECT id
                        FROM {table}
                        WHERE client_path = %s::ltree
                        ORDER BY response_timestamp ASC
                        LIMIT %s
                    )
                    RETURNING id
                """).format(table=table_ident)

                self.cursor.execute(delete_query, (client_path, records_to_remove))
                path_result['removed'] = len(self.cursor.fetchall())
            
            # Add missing records
            elif current_count < queue_length:
                records_to_add = queue_length - current_count
                insert_query = sql.SQL("""
                    INSERT INTO {table} (
                        request_id, client_path, server_path,
                        transaction_tag, rpc_action,
                        response_payload, response_timestamp, is_new_result
                    )
                    VALUES (%s, %s::ltree, %s::ltree, %s, %s, %s::jsonb, NOW(), FALSE)
                """).format(table=table_ident)

                for _ in range(records_to_add):
                    self.cursor.execute(insert_query, (
                        str(uuid.uuid4()),
                        client_path,
                        client_path,              # default server_path = client_path
                        'none',                   # default transaction_tag
                        'none',                   # default rpc_action
                        json.dumps({})            # empty JSONB payload
                    ))
                    path_result['added'] += 1
            
            results[client_path] = path_result

        return results
                
    def restore_default_values(self):
        """
        Restore default values for all fields in rpc_client_table except for client_path.
        
        This method will:
        1. Generate a unique UUID for request_id for each record
        2. Set server_path to match client_path
        3. Set transaction_tag to 'none'
        4. Set rpc_action to 'none'
        5. Set response_payload to an empty JSON object
        6. Set response_timestamp to current time
        7. Set is_new_result to FALSE
        
        Returns:
            int: Number of records updated
        """
        # Update all records with default values while preserving client_path
        # Each record gets a unique UUID through the subquery
        update_query = sql.SQL("""
            UPDATE {table_name}
            SET 
                request_id = (SELECT gen_random_uuid()),  -- Unique UUID per record
                server_path = client_path,  -- Set server_path to match client_path
                transaction_tag = 'none',
                rpc_action = 'none',
                response_payload = '{{}}'::jsonb,
                response_timestamp = NOW(),
                is_new_result = FALSE
            
            RETURNING id
        """).format(table_name=sql.Identifier(self.table_name))
        
        self.cursor.execute(update_query)
        
        # Get count of updated records
        updated_count = len(self.cursor.fetchall())
        
        return updated_count
            
    def check_installation(self):     
        
        try:
            query = sql.SQL("""
            SELECT path, properties FROM {table_name} 
            WHERE label = 'KB_RPC_CLIENT_FIELD';
            """).format(table_name=sql.Identifier(self.database))
            
            self.cursor.execute(query)
            specified_paths_data = self.cursor.fetchall()
            
            paths = []
            lengths = []
            print("specified_paths_data", specified_paths_data)
            
            for row in specified_paths_data:
                paths.append(row[0])
                properties = row[1]
                lengths.append(properties['queue_depth'])
            # Create a dictionary with path as key and other fields as a nested dictionary
            
        except Exception as e:
            raise Exception(f"Error retrieving knowledge base fields: {str(e)}")
        
        self.remove_unspecified_entries(paths)
        self.adjust_queue_length(paths, lengths)
        self.restore_default_values()
    